primitives/state-machine/src/overlayed_changes/offchain.rs

```diff
@@ -0,0 +1,254 @@
+
+/// In-memory storage for offchain workers recoding changes for the actual offchain storage implementation.
+#[derive(Debug, Clone)]
+pub enum OffchainOverlayedChanges {
+	/// Writing overlay changes to the offchain worker database is disabled by configuration.
+	Disabled,
+	/// Overlay changes can be recorded using the inner collection of this variant,
+	/// where the identifier is the tuple of `(prefix, key)`.
+	Enabled(OverlayedMap<(Vec<u8>, Vec<u8>), OffchainOverlayedChange>),
+}
+
+impl Default for OffchainOverlayedChanges {
+	fn default() -> Self {
+		Self::Disabled
+	}
+}
+
+impl OffchainOverlayedChanges {
+	/// Create the disabled variant.
+	pub fn disabled() -> Self {
+		Self::Disabled
+	}
+
+	/// Create the enabled variant.
+	pub fn enabled() -> Self {
+		Self::Enabled(OverlayedMap::default())
+	}
+
+	/// Consume the offchain storage and iterate over all key value pairs.
+	pub fn into_iter(self) -> OffchainOverlayedChangesIntoIter {
+		OffchainOverlayedChangesIntoIter::new(self)
+	}
+
+	/// Iterate over all key value pairs by reference.
+	pub fn iter<'a>(&'a self) -> OffchainOverlayedChangesIter {
+		OffchainOverlayedChangesIter::new(&self)
+	}
+
+	/// Drain all elements of changeset.
+	pub fn drain(&mut self) -> OffchainOverlayedChangesIntoIter {
+		OffchainOverlayedChangesIntoIter::drain(self)
+	}
+
+	/// Remove a key and its associated value from the offchain database.
+	pub fn remove(&mut self, prefix: &[u8], key: &[u8]) {
+		if let Self::Enabled(storage) = self {
+			let _ = storage.set(
+				(prefix.to_vec(), key.to_vec()),
+				OffchainOverlayedChange::Remove,
+				None,
+			);
+		}
+	}
+
+	/// Set the value associated with a key under a prefix to the value provided.
+	pub fn set(&mut self, prefix: &[u8], key: &[u8], value: &[u8]) {
+		if let Self::Enabled(storage) = self {
+			let _ = storage.set(
+				(prefix.to_vec(), key.to_vec()),
+				OffchainOverlayedChange::SetValue(value.to_vec()),
+				None,
+			);
+		}
+	}
+
+	/// Obtain a associated value to the given key in storage with prefix.
+	pub fn get(&self, prefix: &[u8], key: &[u8]) -> Option<OffchainOverlayedChange> {
+		if let Self::Enabled(storage) = self {
+			let key = (prefix.to_vec(), key.to_vec());
+			storage.get(&key).map(|entry| entry.value_ref()).cloned()
+		} else {
+			None
+		}
+	}
+
+	/// Reference to inner change set.
+	pub fn overlay(&self) -> Option<&OverlayedMap<(Vec<u8>, Vec<u8>), OffchainOverlayedChange>> {
+		if let Self::Enabled(storage) = self {
+			Some(&storage)
+		} else {
+			None
+		}
+	}
+
+	/// Mutable reference to inner change set.
+	pub fn overlay_mut(&mut self) -> Option<&mut OverlayedMap<(Vec<u8>, Vec<u8>), OffchainOverlayedChange>> {
+		if let Self::Enabled(storage) = self {
+			Some(storage)
+		} else {
+			None
+		}
+	}
+
+}
+
+use sp_std::collections::btree_map;
+
+/// Iterate by reference over the prepared offchain worker storage changes.
+pub struct OffchainOverlayedChangesIter<'i> {
+	inner: Option<sp_std::iter::Map<
```

Comment:
Can't you use something like `impl Iterator<...>`?

Reply:
I wish I could. In fact if I remove all OffchainIterator struct and just have a function return `impl Iterator<...>` it will work nice and easy.
I wanted to keep those iterator types to avoid breaking too much the api, but I guess I will try to remove them, I am not sure they are usefull here.